source/extensions/transport_sockets/tap/tap.h

```diff
@@ -1,22 +1,62 @@
 #pragma once
 
-#include <fstream>
-
 #include "envoy/config/transport_socket/tap/v2alpha/tap.pb.h"
-#include "envoy/data/tap/v2alpha/transport.pb.h"
+#include "envoy/data/tap/v2alpha/wrapper.pb.h"
 #include "envoy/event/timer.h"
 #include "envoy/network/transport_socket.h"
 
+#include "extensions/common/tap/extension_config_base.h"
+#include "extensions/common/tap/tap_config_base.h"
+#include "extensions/transport_sockets/tap/tap_config.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace TransportSockets {
 namespace Tap {
 
+class SocketTapConfigImpl;
+using SocketTapConfigImplSharedPtr = std::shared_ptr<SocketTapConfigImpl>;
+
+class PerSocketTapperImpl : public PerSocketTapper {
+public:
+  PerSocketTapperImpl(SocketTapConfigImplSharedPtr config, const Network::Connection& connection);
+
+  // PerSocketTapper
+  void closeSocket(Network::ConnectionEvent event) override;
+  void onRead(absl::string_view data) override;
+  void onWrite(absl::string_view data, bool end_stream) override;
+
+private:
+  SocketTapConfigImplSharedPtr config_;
+  const Network::Connection& connection_;
+  std::vector<bool> statuses_;
+  // TODO(mattklein123): Buffering the entire trace until socket close won't scale to
+  // long lived connections or large transfers. We could emit multiple tap
+  // files with bounded size, with identical connection ID to allow later
+  // reassembly.
+  std::shared_ptr<envoy::data::tap::v2alpha::BufferedTraceWrapper> trace_;
+};
+
+class SocketTapConfigImpl : public Extensions::Common::Tap::TapConfigBaseImpl,
+                            public SocketTapConfig,
+                            public std::enable_shared_from_this<SocketTapConfigImpl> {
+public:
+  SocketTapConfigImpl(envoy::service::tap::v2alpha::TapConfig&& proto_config,
+                      Extensions::Common::Tap::Sink* admin_streamer, Event::TimeSystem& time_system)
+      : Extensions::Common::Tap::TapConfigBaseImpl(std::move(proto_config), admin_streamer),
+        time_system_(time_system) {}
+
+  // SocketTapConfig
+  PerSocketTapperPtr createPerSocketTapper(const Network::Connection& connection) override {
+    return std::make_unique<PerSocketTapperImpl>(shared_from_this(), connection);
```

Comment:
it is because we need to keep per-socket, etc. state, and have access to the underlying config to perform static match updates with. Happy to alter if you can suggest a better approach though.

Reply:
> it is because we need to keep per-socket, etc. state, and have access to the underlying config to perform static match updates with. 

I understand this part, just curious why the interface/impl (SocketTapConfig/Impl) abstraction is needed as no other subclass is there. Is more stuff coming?

Reply:
It's only needed to ease testing, though I didn't add any tests in this PR. It's more relevant right away if the config classes are split out into a different file per your other request which I can do. I was just trying to reduce the size of this PR. Will do.